### PR TITLE
feat(new): create private endpoint module

### LINKED ISSUE #258: [Module Proposal]: Private Endpoint Library
### Have you checked this module does not already exist in the registry?

Yes

### Why is the module needed?

I can foresee that we will end up with many modules duplicating the same boilerplate code, that is required for common integrations, such as Private Endpoints for Azure Resources that support Private Link Service.

e.g: https://github.com/Azure/bicep-registry-modules/blob/main/modules/compute/container-registry/.bicep/nested_privateEndpoint.bicep

We should look to modularise this and reduce the duplication. Not currently many base modules in play, but just thinking ahead before we do end up with many that might beed to be refactored. We currently have a large collection of private modules (that we are looking to move away from) and we did use this approach and it was helpful in avoiding duplication and reducing overhead when we did, from time to time, need to tweak the pattern for private endpoint support in modules.

### Module path

lib/private-endpoint

### Describe the module

Essentially shift the common Private Endpoint code into its own library module and then modules that require it simply reference it from mcr.microsoft.com.

Advantage here would be that when we have many modules down the road using the same Private Endpoint code, we only have to modify the code in one place if enhancement is required and then update the Bicep registry module reference when modifying the individual modules down the road (or that bit could even be automated).

### PR DESCRIPTION
**Pull request by thomasriley**
## Description

Create a library module that can be consumed from other Bicep Modules for managing Private Endpoints

This closes #258.

## Adding a new module

- [x] A proposal has been submitted and approved.
- [x] I have included "Closes #{module_proposal_issue_number}" in the PR description.
- [x] I have run `brm validate` locally to verify the module files.
- [ ] I have run deployment tests locally to ensure the module is deployable.

### CODE REVIEW

**Comment by thomasriley** [comment]
I have opened two draft PRs to prepare and visualise how this module will be used.

AKV - a module that does not already support PEP - #361
ACR - a module that already nested this code, and will instead use the BR module - #362

Will hold off on creating any more PRs for now, but once this PR is merged I will continue with PRs on other modules.

**Comment by thllxb** [comment]
I think it makes more sense to move this module to `modules/network/`

File: modules/lib/private-endpoint/main.bicep
Comment (thllxb) [inline modules/lib/private-endpoint/main.bicep:5]: I think it make more sense to just create one pep in a module. It makes the code alot cleaner and easier to maintain.
Users can always use `for` loop on the module.

Comment (thllxb) [inline modules/lib/private-endpoint/main.bicep:5]: It would be better if we can use User-defined data types which is now available with bicep 0.17.1

Comment (thomasriley) [reply modules/lib/private-endpoint/main.bicep:5]: Intention for this module is to cut down on duplicate code across all the other base modules, so moving that `for` logic into each module deviates away from that goal.

Comment (thomasriley) [reply modules/lib/private-endpoint/main.bicep]: I have refactored the module in this direction now.

File: modules/network/private-endpoint/main.bicep
Comment (thllxb) [inline modules/network/private-endpoint/main.bicep:63]: this line can be `privateDnsZone.?name ?? 'default'`

File: modules/network/private-endpoint/main.bicep
Comment (thllxb) [inline modules/network/private-endpoint/main.bicep:58]: this resource can use a for loop, like `[ for zone in privateDnsZones: { `

Comment (thllxb) [reply modules/network/private-endpoint/main.bicep:58]: nvm, the loop is in the resource and that is correct.

File: modules/network/private-endpoint/main.bicep
Comment (thllxb) [inline modules/network/private-endpoint/main.bicep:26]: `param tags {*:string} = {}` is better

File: modules/network/private-endpoint/main.bicep
Comment (thllxb) [inline modules/network/private-endpoint/main.bicep:5]: it can use a default as `resourceGroup().location`, as per https://github.com/Azure/bicep-registry-modules/blob/main/CONTRIBUTING.md#parameters

File: modules/network/private-endpoint/main.bicep
Comment (thllxb) [inline modules/network/private-endpoint/main.bicep:2]:
```suggestion
@minLength(2)
@maxLength(64)
param name string
```

File: modules/network/private-endpoint/main.bicep
Comment (thllxb) [reply modules/network/private-endpoint/main.bicep:26]:
```suggestion
param tags {*:string} = {} 
```

File: modules/network/private-endpoint/main.bicep
Comment (thllxb) [reply modules/network/private-endpoint/main.bicep:63]:
```suggestion
      name: privateDnsZone.?name ?? 'default'
```

File: modules/network/private-endpoint/main.bicep
Comment (thllxb) [reply modules/network/private-endpoint/main.bicep:29]:
```suggestion
  name: name
```

**Review by skydiver1980 — APPROVED** [review]

**Comment by ChrisSidebotham** [comment]
Dear @thomasriley , Azure Verified Modules ([AVM](https://aka.ms/AVM)) will become the single Microsoft standard for Bicep modules. Transitioning to AVM, requests for new features or bug fixes will need to be submitted against the appropriate AVM module via [AVM Module Issues](https://aka.ms/AVM/Bicep/ModuleIssue). 

For bugs or new features with existing modules in the `/modules` directory of this repository, we encourage you to first try the AVM equivalent module if available and if the module is lacking any features, you can file an [AVM Module Issue](https://aka.ms/AVM/Bicep/ModuleIssue). If the module is not available yet, you can file a new [AVM Module Proposal](https://aka.ms/AVM/ModuleProposal).

See this [informational notice](https://github.com/Azure/bicep-registry-modules?tab=readme-ov-file#%EF%B8%8F-upcoming-changes-%EF%B8%8F) for more details on the upcoming changes.

Thank you for your understanding!

**Pull request closed without merging**
